core/api/daemon/src/mill/api/daemon/MillURLClassLoader.scala

```diff
@@ -4,19 +4,89 @@ import java.net.URLClassLoader
 
 /**
  * Convenience wrapper around `java.net.URLClassLoader`. Allows configuration
- * of shared class prefixes, and tracks creation/closing to help detect leaks
+ * of shared class prefixes, and tracks creation/closing to help detect leaks.
+ *
+ * Intrinsic: This classloader is parallel capable.
+ * Due to the way parallel capable classloader need to be registered in the JVM,
+ * we need to drop and never use the first class instance that was created.
+ *
+ * In Java, parallel capable classloaders need to invoke a caller sensitive
+ * protected static method. Since Scala has no `static` initialization syntax,
+ * the only way to invoke a caller-sensitive method is from within the class (instance).
+ * Since this instance isn't yet seen as parallel capable from the JVM, we need to
+ * drop it and create a new instance. All that logic is nicely hidden from the user
+ * by making this constructor private. Instead, the companion object's
+ * [[MillURLClassLoader.apply()]] method is used to create an instance, which
+ * handles all the initialization logic.
+ *
+ * @constructor
+ *
+ * This private constructor should only be used by the companion object.
+ *
+ * @throws DoNotUseThisInstanceException When the [[initParallelClassloading]] parameter is `true`.
+ *
+ * @param initParallelClassloading When true, this classloader will be registered as
+ *                                 parallel capable by calling the caller sensitive
+ *                                 [[java.lang.ClassLoader.registerAsParallelCapable()]].
+ *                                 The contructor will then throw a
+ *                                 [[MillURLClassLoader.DoNotUseThisInstanceException]] to prevent
+ *                                 any use of this improperly initialized class instance.
  */
-class MillURLClassLoader(
+class MillURLClassLoader private (
     classPath: Iterable[java.nio.file.Path],
     parent: ClassLoader,
     sharedLoader: ClassLoader,
     sharedPrefixes: Iterable[String],
-    val label: String
+    val label: String,
+    initParallelClassloading: Boolean
 ) extends URLClassLoader(
       classPath.iterator.map(_.toUri.toURL).toArray,
       MillURLClassLoader.refinePlatformParent(parent)
     ) {
+
   import MillURLClassLoader.*
+
+  if (initParallelClassloading) {
+    // This method is caller sensitive, so we must call it from inside the class.
+    // The just created class instance is improperly initialized, hence we need to not use it.
+    // Therefore, we throw an exception.
+    java.lang.ClassLoader.registerAsParallelCapable()
+    parallelRegistered = true
+    throw new DoNotUseThisInstanceException()
+  }
+
```

Comment:
Could a `@static` annotation be used here? https://docs.scala-lang.org/sips/static-members.html

Reply:
that would let us create a static method on the `MillURLClassLoader` companion object to call `registerAsParallelCapable` without needing to create an instance of `MillURLClassLoader`

Reply:
According to the SIP it works on methods, and makes the annotated object method a static method in the companion class

> Intuitively, the presence of the `@static` annotation ensures that a field/method is declared as a static member of the companion class